### ISSUE TITLE
Some vars are missing ns metadata.

### ISSUE THREAD

**Issue by borkdude** [babashka/babashka#957]
By @cursive-ide on Slack:

However, in bb for a bunch of vars `(#'global-hierarchy #'*ns* #'*print-meta* #'*out* #'alength #'*math-context* #'*command-line-args* #'*err* #'*print-level* #'*in* #'*print-length* #'*print-namespace-maps* #'*warn-on-reflection* #'*data-readers* #'*file* #'clojure.core/unquote)` I get:
```
user=> (meta #'*ns*)
{:dynamic true}
```

Find them all:

```
user=> (clojure.pprint/pprint (for [ns (all-ns), [k v] (ns-publics ns) :when (not (:ns (meta v)))] [(str ns) k]))
(["babashka.signal" pipe-signal-received?]
 ["clojure.java.shell" with-sh-dir]
 ["clojure.java.shell" with-sh-env]
 ["user" *input*]
 ["babashka.wait" wait-for-port]
 ["babashka.wait" wait-for-path]
 ["clojure.core.server" repl]
 ["clojure.core.server" prepl]
 ["clojure.core.server" io-prepl]
 ["clojure.core.server" start-server]
 ["clojure.stacktrace" root-cause]
 ["clojure.stacktrace" print-trace-element]
 ["clojure.stacktrace" print-throwable]
 ["clojure.stacktrace" print-stack-trace]
 ["clojure.stacktrace" print-cause-trace]
 ["clojure.test" test-all-vars]
 ["clojure.test" do-report]
 ["clojure.test" run-all-tests]
 ["clojure.test" assert-any]
 ["clojure.test" testing-contexts-str]
 ["clojure.test" join-fixtures]
 ["clojure.test" assert-expr]
 ["clojure.test" compose-fixtures]
 ["clojure.test" assert-predicate]
 ["clojure.test" function?]
 ["clojure.test" test-vars]
 ["clojure.test" successful?]
 ["clojure.test" use-fixtures]
 ["clojure.test" inc-report-counter]
 ["clojure.test" testing-vars-str]
 ["clojure.test" test-ns]
 ["clojure.test" run-tests]
 ["clojure.core.protocols" Datafiable]
 ["clojure.core.protocols" Navigable]
 ["clojure.java.browse" *open-url-script*]
 ["clojure.java.browse" browse-url]
 ["clojure.data.csv" read-csv]
 ["clojure.data.csv" write-csv]
 ["clojure.core" macroexpand]
 ["clojure.core" when-first]
 ["clojure.core" find-ns]
 ["clojure.core" cond->>]
 ["clojure.core" multi-fn-impl]
 ["clojure.core" while]
 ["clojure.core" proxy*]
 ["clojure.core" satisfies?]
 ["clojure.core" *print-namespace-maps*]
 ["clojure.core" ns-unmap]
 ["clojure.core" newline]
 ["clojure.core" multi-fn-add-method-impl]
 ["clojure.core" bound-fn]
 ["clojure.core" pr]
 ["clojure.core" push-thread-bindings]
 ["clojure.core" read+string]
 ["clojure.core" remove-ns]
 ["clojure.core" find-var]
 ["clojure.core" reify*]
 ["clojure.core" future-call]
 ["clojure.core" ns-publics]
 ["clojure.core" isa?]
 ["clojure.core" ..]
 ["clojure.core" delay]
 ["clojure.core" swap-vals!]
 ["clojure.core" *command-line-args*]
 ["clojure.core" with-bindings]
 ["clojure.core" future-done?]
 ["clojure.core" multi-fn?-impl]
 ["clojure.core" -reified-methods]
 ["clojure.core" eval]
 ["clojure.core" refer]
 ["clojure.core" if-not]
 ["clojure.core" *print-level*]
 ["clojure.core" alength]
 ["clojure.core" global-hierarchy]
 ["clojure.core" pmap]
 ["clojure.core" when-let]
 ["clojure.core" flush]
 ["clojure.core" if-some]
 ["clojure.core" reset-vals!]
 ["clojure.core" *print-length*]
 ["clojure.core" *file*]
 ["clojure.core" compare-and-set!]
 ["clojure.core" -new-dynamic-var]
 ["clojure.core" pop-thread-bindings]
 ["clojure.core" printf]
 ["clojure.core" ->]
 ["clojure.core" *err*]
 ["clojure.core" doto]
 ["clojure.core" future]
 ["clojure.core" ns-aliases]
 ["clojure.core" read]
 ["clojure.core" ns-resolve]
 ["clojure.core" as->]
 ["clojure.core" when-not]
 ["clojure.core" *print-meta*]
 ["clojure.core" when]
 ["clojure.core" ns-refers]
 ["clojure.core" prn]
 ["clojure.core" extend]
 ["clojure.core" some->>]
 ["clojure.core" default-data-readers]
 ["clojure.core" extends?]
 ["clojure.core" reset-thread-binding-frame-impl]
 ["clojure.core" *math-context*]
 ["clojure.core" use]
 ["clojure.core" unquote]
 ["clojure.core" reset!]
 ["clojure.core" println]
 ["clojure.core" extend-type]
 ["clojure.core" macroexpand-1]
 ["clojure.core" defmethod]
 ["clojure.core" requiring-resolve]
 ["clojure.core" time]
 ["clojure.core" future?]
 ["clojure.core" require]
 ["clojure.core" future-cancelled?]
 ["clojure.core" get-thread-binding-frame-impl]
 ["clojure.core" *data-readers*]
 ["clojure.core" extend-protocol]
 ["clojure.core" cond->]
 ["clojure.core" dotimes]
 ["clojure.core" -new-var]
 ["clojure.core" reify]
 ["clojure.core" instance?]
 ["clojure.core" with-open]
 ["clojure.core" defonce]
 ["clojure.core" the-ns]
 ["clojure.core" record?]
 ["clojure.core" ns-name]
 ["clojure.core" defn-]
 ["clojure.core" *out*]
 ["clojure.core" ns-map]
 ["clojure.core" defprotocol]
 ["clojure.core" swap!]
 ["clojure.core" *warn-on-reflection*]
 ["clojure.core" assert]
 ["clojure.core" print]
 ["clojure.core" *in*]
 ["clojure.core" letfn]
 ["clojure.core" read-line]
 ["clojure.core" binding-conveyor-fn]
 ["clojure.core" descendants]
 ["clojure.core" with-out-str]
 ["clojure.core" condp]
 ["clojure.core" derive]
 ["clojure.core" load-string]
 ["clojure.core" ancestors]
 ["clojure.core" cond]
 ["clojure.core" intern]
 ["clojure.core" with-in-str]
 ["clojure.core" some->]
 ["clojure.core" ns-interns]
 ["clojure.core" all-ns]
 ["clojure.core" binding]
 ["clojure.core" with-local-vars]
 ["clojure.core" ns-imports]
 ["clojure.core" load-file]
 ["clojure.core" alias]
 ["clojure.core" read-string]
 ["clojure.core" proxy]
 ["clojure.core" with-redefs]
 ["clojure.core" locking]
 ["clojure.core" *ns*]
 ["clojure.core" defmulti]
 ["clojure.core" if-let]
 ["clojure.core" underive]
 ["clojure.core" lazy-cat]
 ["clojure.core" comment]
 ["clojure.core" parents]
 ["clojure.core" deref]
 ["clojure.core" resolve]
 ["clojure.core" defrecord]
 ["clojure.core" with-redefs-fn]
 ["clojure.core" protocol-type-impl]
 ["clojure.core" when-some]
 ["clojure.core" ->>]
 ["clojure.core" future-cancel]
 ["clojure.core" refer-clojure]
 ["clojure.core.async" reduce]
 ["clojure.core.async" remove>]
 ["clojure.core.async" timeout]
 ["clojure.core.async" mix]
 ["clojure.core.async" pub]
 ["clojure.core.async" take]
 ["clojure.core.async" <!]
 ["clojure.core.async" map]
 ["clojure.core.async" mapcat>]
 ["clojure.core.async" buffer]
 ["clojure.core.async" close!]
 ["clojure.core.async" offer!]
 ["clojure.core.async" chan]
 ["clojure.core.async" tap]
 ["clojure.core.async" go]
 ["clojure.core.async" admix]
 ["clojure.core.async" promise-chan]
 ["clojure.core.async" unique]
 ["clojure.core.async" solo-mode]
 ["clojure.core.async" transduce]
 ["clojure.core.async" onto-chan]
 ["clojure.core.async" to-chan]
 ["clojure.core.async" dropping-buffer]
 ["clojure.core.async" untap-all]
 ["clojure.core.async" into]
 ["clojure.core.async" >!!]
 ["clojure.core.async" alt!]
 ["clojure.core.async" pipeline]
 ["clojure.core.async" sub]
 ["clojure.core.async" go-loop]
 ["clojure.core.async" map>]
 ["clojure.core.async" pipe]
 ["clojure.core.async" unmix]
 ["clojure.core.async" filter<]
 ["clojure.core.async" remove<]
 ["clojure.core.async" alt!!]
 ["clojure.core.async" <!!]
 ["clojure.core.async" toggle]
 ["clojure.core.async" sliding-buffer]
 ["clojure.core.async" partition]
 ["clojure.core.async" merge]
 ["clojure.core.async" partition-by]
 ["clojure.core.async" unsub-all]
 ["clojure.core.async" >!]
 ["clojure.core.async" split]
 ["clojure.core.async" unmix-all]
 ["clojure.core.async" filter>]
 ["clojure.core.async" untap]
 ["clojure.core.async" alts!]
 ["clojure.core.async" unsub]
 ["clojure.core.async" poll!]
 ["clojure.core.async" thread]
 ["clojure.core.async" map<]
 ["clojure.core.async" do-alts]
 ["clojure.core.async" pipeline-blocking]
 ["clojure.core.async" pipeline-async]
 ["clojure.core.async" mult]
 ["clojure.core.async" thread-call]
 ["clojure.core.async" alts!!]
 ["clojure.core.async" mapcat<]
 ["clojure.core.async" unblocking-buffer?]
 ["clojure.core.async" do-alt]
 ["clojure.core.async" put!]
 ["clojure.core.async" take!]
 ["clojure.main" demunge]
 ["clojure.main" repl-requires]
 ["clojure.main" repl]
 ["clojure.core.async.impl.protocols" ReadPort]
 ["clojure.java.io" as-file]
 ["clojure.java.io" Coercions]
 ["clojure.java.io" as-url]
 ["clojure.repl" source-fn]
 ["clojure.repl" doc]
 ["clojure.repl" stack-element-str]
 ["clojure.repl" find-doc]
 ["clojure.repl" dir]
 ["clojure.repl" pst]
 ["clojure.repl" dir-fn]
 ["clojure.repl" source]
 ["clojure.repl" demunge]
 ["clojure.repl" apropos]
 ["clojure.lang" resetVals]
 ["clojure.lang" reset]
 ["clojure.lang" swapVals]
 ["clojure.lang" compareAndSet]
 ["clojure.lang" swap]
 ["clojure.lang" deref]
 ["clojure.template" apply-template]
 ["clojure.template" do-template])
```

**Comment by bobisageek** [comment]
I started tinkering with this a little bit, basically just a function to get some of the meta from a clojure var and merge in extra meta so I could experiment with the first few things. Getting the meta works (although like you said, it'd be a bit tedious to handle them), but in namespace creation cases where the symbol gets mapped to a function, `doc` doesn't work because the mapped thing isn't a Var, for example:

https://github.com/babashka/babashka/blob/89695d8f7a5b4d46e3b6105a625d66238d5d3178/src/babashka/impl/clojure/java/shell.clj#L41-L45

We can tack some of the meta from clojure.java.shell/with-sh-dir onto that, and then in bb, `meta` will work, but `doc` won't, unless we also make it a Var with `(sci/new-var ...)`.

I realize that making `doc` work is not a stated goal, but I also thought that if there was to be a big effort to add the metadata, then maybe it makes sense to tack on a little bit more work to add some extra value (assuming that `doc` working is a value add).

I also realize you're already working on this, so you may have already had all these thoughts, so if you just want some help grinding a pattern you've come up with over that big list of vars, I can put some time in over the weekend to help out with that.

**Comment by borkdude** [comment]
Hi Bob, thanks for the offer. I have a branch called repair-metadata (or so, I’m typing from my phone). I’ll just merge that to master and you can pick up where I left.

**Comment by borkdude** [comment]
I’ll merge in a couple of hours but feel free to look at the branch meanwhile. A lot of those vars are actually in SCI as well.

**Comment by borkdude** [comment]
@bobisageek ok, I merged the work I had done on this to sci master + bb master. Feel free to look at those PRs as inspiration. Also feel free to PR the changes in smaller chunks at a time, it doesn't have to be one big bang PR.

**Comment by borkdude** [comment]
@bobisageek has solved most of the problems, thanks much. 

The `clojure.lang` namespace is a bit of a hack which I will find better solution for hopefully and once that's solved we can close this and there will be 0 vars without ns metadata in bb. The `*input*` variable we can make private I think, since it's only intended to be used from the `user` namespace on the command line.

**Comment by borkdude** [comment]
I will get back to clojure.lang in another issue, for now I think this is sufficiently handled? Let me know if I should re-open.
